macaroon.js

```diff
@@ -2,30 +2,267 @@
+
+/**
+ * Convert a string to a Uint8Array by utf-8
+ * encoding it.
+ * @param {string} s The string to convert.
+ * @return {Uint8Array} The resulting bytes.
+ */
+const stringToBytes = function(s) {
+  if (s === null) {
+    return null;
+  }
+  return utf8Encoder.encode(s);
 };
 
-// Uses protobuf-style varints
-function toVarintBuf(number) {
-  return Buffer.from(varint.encode(number));
-}
+/**
+ * Convert a Uint8Array to a string by
+ * utf-8 decoding it. Throws an exception if
+ * the bytes do not represent well-formed utf-8.
+ * @param {Uint8Array} The bytes to convert.
+ * @return {string} The resulting string.
+ */
+const bytesToString = function(b) {
+  if (b === null) {
+    return null;
+  }
+  return utf8Decoder.decode(b);
+};
+
+/**
+ * The maximum integer that can be manipulated with
+ * JS bitwise operations.
+ */
+const maxInt = Math.pow(2, 32)-1;
+
+const ByteBuffer = class ByteBuffer {
+    /**
+     * Create a new ByteBuffer. A ByteBuffer holds
+     * a Uint8Array that it grows when written to.
+     * @param {int} The initial capacity of the buffer.
+     */
+  constructor(capacity) {
+    this._buf = new Uint8Array(capacity);
+    this._length = 0;
+  }
+  /**
+   * Append several bytes to the buffer.
+   * @param {Uint8Array} The bytes to append.
+   */
+  appendBytes(bytes) {
+    this._grow(this._length + bytes.length);
+    this._buf.set(bytes, this._length);
+    this._length += bytes.length;
+  }
+  /**
+   * Append a single byte to the buffer.
+   * @param {int} The byte to append
+   */
+  appendByte(byte) {
+    this._grow(this._length + 1);
+    this._buf[this._length] = byte;
+    this._length++;
+  }
+  /**
+   * Append a variable length integer to the buffer.
+   * @param {int} The number to append.
+   */
+  appendUvarint(x) {
+    if (x > maxInt || x < 0) {
+      throw new RangeError(`varint ${x} out of range`);
+    }
+    this._grow(this._length + maxVarintLen32);
+    var i = this._length;
+    while(x >= 0x80) {
+      this._buf[i++] = (x & 0xff) | 0x80;
+      x >>>= 7;
+    }
+    this._buf[i++] = x | 0;
+    this._length = i;
+  }
+  /**
+   * Return everything that has been appended to the buffer.
+   * Note that the returned array is shared with the internal buffer.
+   * @return {Uint8Array} The buffer.
+   */
+  get bytes() {
+    return this._buf.subarray(0, this._length);
+  }
+  /**
+   * Grow the internal buffer so that it's at least as but as minCap.
+   * @param {int} The minimum new capacity.
+   */
+  _grow(minCap) {
+    if (minCap <= this._capacity) {
+      return;
+    }
+    // TODO could use more intelligent logic to grow more slowly on large buffers.
+    const doubleCap = this._buf.length * 2;
+    const newCap = minCap > doubleCap ? minCap : doubleCap;
+    const newContent = new Uint8Array(newCap);
+    newContent.set(this._buf.subarray(0, this._length));
+    this._buf = newContent;
+  }
+};
+
+const maxVarintLen32 = 5;
+
+const ByteReader = class byteReader {
```

Comment:
`class byteReader` -> `class ByteReader`
Classes *should* be capitalised.

Reply:
done